File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        wordSet = set(wordDict)
        maxLen = max(map(len, wordDict))
        dp = [False] * (n+1)
        dp[0] = True
        for i in range(1,n+1):
            for len_ in range(1,min(i,maxLen)+1):
                j=i-len_
                if dp[j] and s[j:j+len_] in wordSet:
                    dp[i] = True
                    break
        return dp[n]

File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet(wordDict.begin(),wordDict.end());
        int maxLen = 0;
        for (const string& word : wordDict)
            maxLen = max(maxLen, (int)word.size());
        int n = (int)s.size(); // length of string
        vector<bool> dp(n+1,false);
        dp[0] = true;
        for (int i=1; i<=n; i++){
            for (int len=1; len<=min(i,maxLen); len++){
                int j = i-len;
                dp[i] = dp[j] && wordSet.contains(s.substr(j,len));
                if (dp[i])
                    break;
            }
        }
        return dp[n];
    }
};